callbacks.py

```diff
@@ -91,3 +97,99 @@ def get_prediction_masks(self):
+
+    def _get_validation_loss(self):
+        outputs = self._transform()
+        prediction = self._generate_prediction(outputs)
+        if len(prediction) == 0:
+            return self.validation_loss.setdefault(self.epoch_id, {'sum': Variable(torch.Tensor([0]))})
+
+        with TemporaryDirectory() as temp_dir:
+            prediction_filepath = os.path.join(temp_dir, 'prediction.json')
+            with open(prediction_filepath, "w") as fp:
+                fp.write(json.dumps(prediction))
+
+            annotation_file_path = os.path.join(self.data_dir, 'val', "annotation.json")
+
+            logger.info('Calculating mean precision and recall')
+            average_precision, average_recall = coco_evaluation(gt_filepath=annotation_file_path,
+                                                                prediction_filepath=prediction_filepath,
+                                                                image_ids=self.meta_valid[Y_COLUMNS_SCORING].values,
+                                                                category_ids=CATEGORY_IDS[1:],
+                                                                small_annotations_size=self.small_annotations_size)
+        return self.validation_loss.setdefault(self.epoch_id, {'sum': Variable(torch.Tensor([average_precision]))})
+
+    def _transform(self):
+        self.model.eval()
+        batch_gen, steps = self.validation_datagen
+        outputs = {}
+        for batch_id, data in enumerate(batch_gen):
+            if isinstance(data, list):
+                X = data[0]
+            else:
+                X = data
+
+            if torch.cuda.is_available():
+                X = Variable(X, volatile=True).cuda()
+            else:
+                X = Variable(X, volatile=True)
+
+            outputs_batch = self.model(X)
+            if len(self.output_names) == 1:
+                outputs.setdefault(self.output_names[0], []).append(outputs_batch.data.cpu().numpy())
+            else:
+                for name, output in zip(self.output_names, outputs_batch):
+                    output_ = output.data.cpu().numpy()
+                    outputs.setdefault(name, []).append(output_)
+            if batch_id == steps:
+                break
+        self.model.train()
+        outputs = {'{}_prediction'.format(name): np.vstack(outputs_) for name, outputs_ in outputs.items()}
+        for name, prediction in outputs.items():
+            outputs[name] = softmax(prediction, axis=1)
+
+        return outputs
+
+    def _generate_prediction(self, outputs):
+        data = {'input': {'meta': self.meta_valid,
+                          'meta_valid': None,
+                          'train_mode': False,
+                          'target_sizes': [(300, 300)] * len(self.meta_valid),
+                          **outputs
+                          },
+                }
+
+        self.validation_pipeline.clean_cache()
+        output = self.validation_pipeline.fit_transform(data)
```

Comment:
@taraspiotr I understand why we are using fit_transform rather than transform here. It's just we are validating so we shouldn't be fitting anything. That means having a flag in Step constructor `is_trainable=False` that makes it possible to create transformers on the fly would clean it up. So for now let's leave it as is just create a ticket to `steppy` with a feature request for it.